spring-cloud-stream/src/test/java/org/springframework/cloud/stream/function/FunctionSupportTests.java

```diff
@@ -0,0 +1,121 @@
+import org.springframework.integration.dsl.IntegrationFlowBuilder;
+import org.springframework.integration.dsl.IntegrationFlows;
+import org.springframework.integration.scheduling.PollerMetadata;
+import org.springframework.integration.support.MessageBuilder;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.scheduling.support.PeriodicTrigger;
+import org.springframework.util.MimeTypeUtils;
+import org.springframework.util.StringUtils;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ *
+ * @author Oleg Zhurakousky
+ *
+ * @since 2.1
+ *
+ */
+public class FunctionSupportTests {
+
+	@Test
+	public void testFunctionIsAppliedToMessageSource() {
+		ApplicationContext context = new SpringApplicationBuilder(TestChannelBinderConfiguration.getCompleteConfiguration(MyFunctionsConfiguration.class))
+				.web(WebApplicationType.NONE).run("--spring.cloud.stream.function.name=toUpperCase",  "--spring.jmx.enabled=false");
+
+		OutputDestination target = context.getBean(OutputDestination.class);
+		assertThat(target.receive(1000).getPayload()).isEqualTo("HELLO FUNCTION".getBytes(StandardCharsets.UTF_8));
+	}
+
+	@Test
+	public void testComposedFunctionIsAppliedToMessageSource() {
+		ApplicationContext context = new SpringApplicationBuilder(TestChannelBinderConfiguration.getCompleteConfiguration(MyFunctionsConfiguration.class))
+				.web(WebApplicationType.NONE).run("--spring.cloud.stream.function.name=toUpperCase|concatWithSelf", "--spring.jmx.enabled=false");
+
+		OutputDestination target = context.getBean(OutputDestination.class);
+		assertThat(target.receive(1000).getPayload()).isEqualTo("HELLO FUNCTION:HELLO FUNCTION".getBytes(StandardCharsets.UTF_8));
+	}
+
+	@SpringBootApplication
+	@Import(SourceConfiguration.class)
+	public static class MyFunctionsConfiguration {
+		@Bean
+		public Function<String, String> toUpperCase() {
+			return x -> x.toUpperCase();
+		}
+
+		@Bean
+		public Function<String, String> concatWithSelf() {
+			return x -> x + ":" + x;
+		}
+	}
+
+	@EnableBinding(Source.class)
+	public static class SourceConfiguration {
+
+		@Autowired
+		private Source source;
+
+		@Autowired
+		private FunctionProperties functionProperties;
+
+		@Bean
+		public IntegrationFlow messageSourceFlow(FunctionSupport<?, ?> functionSupport) {
+			Supplier<Message<String>> messageSource = () -> MessageBuilder.withPayload("hello function")
+					.setHeader(MessageHeaders.CONTENT_TYPE, MimeTypeUtils.TEXT_PLAIN).build();
+
+			PollerMetadata pm = new PollerMetadata();
+			pm.setTrigger(new PeriodicTrigger(1, TimeUnit.SECONDS));
+			pm.setMaxMessagesPerPoll(1);
+			IntegrationFlowBuilder flowBuilder = IntegrationFlows.from(messageSource);
+
+			if (StringUtils.hasText(functionProperties.getName())) {
+				functionSupport.applyFunctionToIntegrationFlow(functionProperties.getName(), flowBuilder, message -> source.output().send(message));
```

Comment:
Well, sure, you can think about it, but remember, this work was retrofitted front the work Stephan and I did with non-SI binders (new binders), so Consumer really morphs into an output Publisher (in reactive terms). Sure in this particular case sending to a channel is something that you may take care of on the SI side, but that would not help me in the long run since I would need to have separate support classes for SI and non-SI binders.